### ISSUE TITLE
Volatile functions for signup and change_password?

### ISSUE THREAD

**Issue by richcorbs** [sander-io/hasura-jwt-auth#5]
This is an awesome approach. I love it and wish it were part of Hasura main.

When VOLATILE stored procedures are available (soon I hope) will you add support for signup and change_password functions?

**Comment by richcorbs** [comment]
Wow, now I see the encrypt password trigger! Clever.

This is even more amazing than I thought. Thank you for this!

**Comment by sanderhahn** [comment]
Thank you for your kind words :) Of course its the flexible design of Hasura and PostgreSQL that makes it possible! Will see if i can update the code to support password change once volatile is supported. https://github.com/hasura/graphql-engine/issues/1514

**Comment by richcorbs** [comment]
Is the "before save trigger" a workaround until volatile functions are available?

**Comment by sanderhahn** [comment]
Think that at the moment the limitation is: generating a random password on the server and returning that value while also storing it encrypted. This is also necessary to make something like `access_tokens` work, have not figured out how to do this properly. 

Think that for both signup and password reset you will need some external service to present a form and send a mail with a verification token. Once you are already using an external service it might be more convenient to have that also perform these manipulations using direct sql connection or a special Hasura role.

**Comment by sanderhahn** [comment]
Signup and password reset flow could be something like this:

- `signup(email, password)` -> sends verification_link with verification token in email
- `verify(verification_token)` -> enables user on system
- `request_password_reset(email)` -> sends password_reset_link with verification token in email
- `password_reset(verification_token, new_password)` -> changes password of user on system
- `invite(email)` -> sends a signup link in email

However think that it requires an email service and a website that hosts the signup form/change password form and verification actions.

**Comment by richcorbs** [comment]
Yep I can see that. I was thinking more simply and probably used the wrong
words. I meant the ability for the user to change their password and create a password which the trigger solves.

The points you make are good and valid and not something I'd personally
expect/need to have out of the box starting a new project.

**Comment by sanderhahn** [comment]
At the moment its possible for users to update their own email/password using:

```graphql
mutation {
  update_hasura_user(_set: {cleartext_password: "changed_password"}, where: {}) {
    returning {
      id
    }
  }
}
```

Tested this using the metadata from the Quickstart, that allows updating the email and the password for the currently active user.

![User update permissions 2019-08-27 12-08-07](https://user-images.githubusercontent.com/775103/63762591-63cbf680-c8c3-11e9-870c-60312e565aa9.png)

**Comment by cpursley** [comment]
As @richcorbs mentioned, this is an awesome approach.

Any thoughts on how to do account verification?

My immediate thought was on the user table, add two columns: `verification_token` and `verified`. The verification_token can be a uuid that gets generation on record creation and verified a boolean that by default is _false_.

Then have a `verify_hasura_user` Postgres function and corresponding Hasura mutation that looks up the user by the verification_token and then sets the verified column to true. Modify the hasura_auth function to fail if verified field has false value.

A url containing the GraphQL mutation along with the verification_token would be sent to the user to click in an email.

Thoughts?

**Comment by richcorbs** [comment]
Seems legit to me. I use a similar method for password resets.

**Comment by cpursley** [comment]
@richcorbs ~ how did you get around the VOLATILE restriction?

**Comment by richcorbs** [comment]
Ah, right. I'm not using a function for password resets.

**Comment by cpursley** [comment]
I ended up with something like this:

```
create or replace function account_verify() returns trigger as $$
begin
    if old.verified is not true and
       new.verify_token is not null and
       new.verify_token = old.verification_token
       then
		new.verify_token = '';
		new.verified = true;
    end if;
	new.verify_token = '';
    return new;
end;
$$ language 'plpgsql';

create trigger account_verify_trigger
before update of verify_token on user_account
for each row execute procedure account_verify();
```

Where the hasura update account function (same as used for changing password) takes the verify_token param to match against the verification_token param via trigger. I'm also now checking if verified = true in the hasura_auth function to prevent unverified users from getting a jwt token.

Any ideas on how to beef this up ~ make it more secure / reliable?
